File: frontend/src/features/company/orderProfile.js
export const EMPTY_ORDER_PROFILE = {
  legal_name: "",
  consignee: "",
  address: "",
  contact_name: "",
  contact_phone: "",
  carrier: "",
  delivery_payer: "",
  delivery_destination: "",
  brand_terms: [],
};

export const ORDER_PROFILE_BRANDS = [
  { id: "angiopharm", label: "ANGIOPHARM", fields: ["discount", "customer_type", "payment_method", "payment_control"] },
  { id: "skin_synergy", label: "Skin Synergy", fields: ["discount", "dealer_name"] },
  { id: "klapp", label: "KLAPP", fields: ["discount"] },
];

export function normalizeOrderProfile(payload) {
  const source = payload || {};
  return {
    ...EMPTY_ORDER_PROFILE,
    ...source,
    brand_terms: Array.isArray(source.brand_terms) ? source.brand_terms.map((item) => ({ ...item })) : [],
  };
}

export function termsFor(profile, brand) {
  return profile.brand_terms.find((item) => item.brand === brand) || {
    brand,
    dealer_name: "",
    payment_method: "",
    payment_control: "",
    customer_type: "",
    discount_basis_points: 0,
    discount_set: false,
  };
}

export function setTerms(profile, brand, patch) {
  const next = { ...termsFor(profile, brand), ...patch, brand };
  const others = profile.brand_terms.filter((item) => item.brand !== brand);
  return { ...profile, brand_terms: [...others, next] };
}

export function discountText(terms) {
  if (!terms.discount_set) return "";
  return String((Number(terms.discount_basis_points) || 0) / 100).replace(".", ",");
}

export function parseDiscount(value) {
  const text = String(value ?? "").trim();
  if (!text) return { set: false, basisPoints: 0, issue: "" };
  if (!/^\d{1,3}(?:[.,]\d{1,2})?$/.test(text)) {
    return { set: false, basisPoints: 0, issue: "Введите скидку от 0 до 100, не более двух знаков после запятой." };
  }
  const percent = Number(text.replace(",", "."));
  if (!Number.isFinite(percent) || percent < 0 || percent > 100) {
    return { set: false, basisPoints: 0, issue: "Скидка должна быть от 0 до 100%." };
  }
  return { set: true, basisPoints: Math.round(percent * 100), issue: "" };
}

export function orderProfileIssue(profile, discountInputs) {
  const phone = profile.contact_phone.trim();
  if (phone && (!/^[+\d][\d\s()+-]*$/.test(phone) || phone.replace(/\D/g, "").length < 6)) {
    return "Проверьте телефон: нужны минимум 6 цифр, допустимы +, скобки, пробелы и дефисы.";
  }
  for (const brand of ORDER_PROFILE_BRANDS) {
    const parsed = parseDiscount(discountInputs[brand.id]);
    if (parsed.issue) return `${brand.label}: ${parsed.issue}`;
  }
  return "";
}

export function profileForSave(profile, discountInputs) {
  return {
    ...profile,
    brand_terms: ORDER_PROFILE_BRANDS.map((brand) => {
      const parsed = parseDiscount(discountInputs[brand.id]);
      return {
        ...termsFor(profile, brand.id),
        discount_set: parsed.set,
        discount_basis_points: parsed.basisPoints,
      };
    }),
  };
}

File: backend/services/identity-service/internal/domain/order_profile.go
package domain

import (
	"fmt"
	"regexp"
	"strings"
)

const (
	maxProfileTextLength = 300
	maxBrandTerms        = 20
	maxDiscountBPS       = 10_000
)

var brandKeyPattern = regexp.MustCompile(`^[a-z][a-z0-9_]{0,49}$`)
var phonePattern = regexp.MustCompile(`^[+0-9][0-9 ()+-]*$`)

func NormalizeOrderProfile(profile OrderProfile) (OrderProfile, error) {
	profile.LegalName = strings.TrimSpace(profile.LegalName)
	profile.Consignee = strings.TrimSpace(profile.Consignee)
	profile.Address = strings.TrimSpace(profile.Address)
	profile.ContactName = strings.TrimSpace(profile.ContactName)
	profile.ContactPhone = strings.TrimSpace(profile.ContactPhone)
	profile.Carrier = strings.TrimSpace(profile.Carrier)
	profile.DeliveryPayer = strings.TrimSpace(profile.DeliveryPayer)
	profile.DeliveryDestination = strings.TrimSpace(profile.DeliveryDestination)

	for name, value := range map[string]string{
		"legal_name": profile.LegalName, "consignee": profile.Consignee, "address": profile.Address,
		"contact_name": profile.ContactName, "contact_phone": profile.ContactPhone, "carrier": profile.Carrier,
		"delivery_payer": profile.DeliveryPayer, "delivery_destination": profile.DeliveryDestination,
	} {
		if len([]rune(value)) > maxProfileTextLength {
			return OrderProfile{}, fmt.Errorf("%w: %s is too long", ErrInvalid, name)
		}
	}
	if profile.ContactPhone != "" {
		digits := 0
		for _, char := range profile.ContactPhone {
			if char >= '0' && char <= '9' {
				digits++
			}
		}
		if !phonePattern.MatchString(profile.ContactPhone) || digits < 6 {
			return OrderProfile{}, fmt.Errorf("%w: invalid contact_phone", ErrInvalid)
		}
	}
	if len(profile.BrandTerms) > maxBrandTerms {
		return OrderProfile{}, fmt.Errorf("%w: too many brand terms", ErrInvalid)
	}
	seen := make(map[string]struct{}, len(profile.BrandTerms))
	terms := make([]BrandTerms, 0, len(profile.BrandTerms))
	for _, item := range profile.BrandTerms {
		item.Brand = strings.ToLower(strings.TrimSpace(item.Brand))
		item.DealerName = strings.TrimSpace(item.DealerName)
		item.PaymentMethod = strings.TrimSpace(item.PaymentMethod)
		item.PaymentControl = strings.TrimSpace(item.PaymentControl)
		item.CustomerType = strings.TrimSpace(item.CustomerType)
		if !brandKeyPattern.MatchString(item.Brand) {
			return OrderProfile{}, fmt.Errorf("%w: invalid brand", ErrInvalid)
		}
		if _, ok := seen[item.Brand]; ok {
			return OrderProfile{}, fmt.Errorf("%w: duplicate brand", ErrInvalid)
		}
		seen[item.Brand] = struct{}{}
		if !item.DiscountSet {
			item.DiscountBasisPoints = 0
		}
		if item.DiscountBasisPoints < 0 || item.DiscountBasisPoints > maxDiscountBPS {
			return OrderProfile{}, fmt.Errorf("%w: discount must be between 0 and 100", ErrInvalid)
		}
		for _, value := range []string{item.DealerName, item.PaymentMethod, item.PaymentControl, item.CustomerType} {
			if len([]rune(value)) > maxProfileTextLength {
				return OrderProfile{}, fmt.Errorf("%w: brand term is too long", ErrInvalid)
			}
		}
		terms = append(terms, item)
	}
	profile.BrandTerms = terms
	return profile, nil
}
